src/dataset.rs

```diff
@@ -508,6 +511,95 @@ impl Dataset {
         }
         Ok(Transaction::new(self))
     }
+
+    /// Execute a SQL query against the Dataset. It is equivalent to calling
+    /// [`GDALDatasetExecuteSQL`](https://gdal.org/api/raster_c_api.html#_CPPv421GDALDatasetExecuteSQL12GDALDatasetHPKc12OGRGeometryHPKc).
+    /// Returns a `ResultSet`, which can be treated just as any other [`Layer`], or `None` if the
+    /// query does not have a `ResultSet`, such as `ALTER TABLE`, etc. Note that this is different
+    /// than a query which has an *empty* `ResultSet`.
+    ///
+    /// Example
+    /// ---
+    ///
+    /// ```
+    /// # use gdal::Dataset;
+    /// # use std::path::Path;
+    /// #
+    /// let ds = Dataset::open(Path::new("fixtures/roads.geojson")).unwrap();
+    /// let query = "SELECT kind, is_bridge, highway FROM roads WHERE highway = 'pedestrian'";
+    /// let result_set = ds.execute_sql(query, None, None).unwrap().unwrap();
+    ///
+    /// assert_eq!(10, result_set.feature_count());
+    ///
+    /// for feature in result_set.features() {
+    ///     let highway = feature
+    ///         .field("highway")
+    ///         .unwrap()
+    ///         .unwrap()
+    ///         .into_string()
+    ///         .unwrap();
+    ///
+    ///     assert_eq!("pedestrian", highway);
+    /// }
+    /// ```
+    pub fn execute_sql<S: AsRef<str>>(
+        &self,
+        query: S,
+        spatial_filter: Option<&Geometry>,
+        dialect: Option<S>,
+    ) -> Result<Option<ResultSet>> {
+        let query = CString::new(query.as_ref())?;
+
+        let dialect_c_str = match dialect {
+            None => None,
+            Some(d) => Some(CString::new(d.as_ref())?),
+        };
+
+        self._execute_sql(query, spatial_filter, dialect_c_str)
+    }
+
+    fn _execute_sql(
+        &self,
+        query: CString,
+        spatial_filter: Option<&Geometry>,
+        dialect_c_str: Option<CString>,
+    ) -> Result<Option<ResultSet>> {
+        let mut filter_geom: OGRGeometryH = std::ptr::null_mut();
+
+        let dialect_ptr = match dialect_c_str {
+            None => std::ptr::null(),
+            Some(ref d) => d.as_ptr(),
+        };
+
+        if let Some(spatial_filter) = spatial_filter {
+            filter_geom = unsafe { spatial_filter.c_geometry() };
+        }
+
+        let c_dataset = unsafe { self.c_dataset() };
+
+        unsafe { gdal_sys::CPLErrorReset() };
+
+        let c_layer = unsafe {
+            gdal_sys::GDALDatasetExecuteSQL(c_dataset, query.as_ptr(), filter_geom, dialect_ptr)
+        };
+
+        let cpl_err = unsafe { gdal_sys::CPLGetLastErrorType() };
+
+        if cpl_err != CPLErr::CE_None {
+            return Err(_last_cpl_err(cpl_err));
+        }
+
+        if c_layer.is_null() {
```

Comment:
I'm not sure, I was also confused by the vagueness of the docs for this function. My understanding is that it is only null if the SQL statement is one which has no result set, i.e. ALTER TABLE, CREATE INDEX, DROP INDEX, etc.

I'm not sure if we can assume that if the pointer is non-null that there was no error. I suppose it's possible for the GDAL code to throw an error in the middle of executing the SQL statement, and I don't know if it still returns a result set in this case. It seems prudent to check for an error regardless of what is returned.

Reply:
That sentence seems pretty unambiguous to me.

It might be also possible that GDAL encounters an error it can handle, uses a fallback path and returns a valid result, but doesn't reset the last error. But let's leave it like this, the Python bindings check for errors first.

Who knew that error checking could be so complicated?

Reply:
I mean we are talking about a code base that mixes C, C++, and Rust. Global error state and inconsistent call site checking is pretty par for the course with C lol. That's why I tend to favor defensive code, even at the expense of repetition and readability. In my experience, you can't rely on other code to do the right thing 100% of the time.